### DISCUSSION TITLE
How to bind Mod4+button2 (Win + middle-click) to close the window?

### DISCUSSION THREAD

**Discussion by CyberShadow** [i3/i3#4511] [Q&A]
Attempt 1:
```
bindsym --border --whole-window           Mod4+button2 exec "i3-msg kill"
```
It works, however, when I release the middle mouse button, that goes to whichever window is now under the mouse button (e.g. for terminals it pastes the selection).

Attempt 2:
```
bindsym --border --whole-window           Mod4+button2 exec "i3-msg kill"
bindsym --border --whole-window --release Mod4+button2 nop
```
Doesn't work (same effect as above).

Attempt 3:
```
bindsym --border --whole-window --release Mod4+button2 exec "i3-msg kill"
```
Doesn't work (middle click does nothing but go through to the application).

Looking at the log, I see that i3 never even receives XCB_BUTTON_RELEASE. Is it some kind of Xorg limitation? I saw this also mentioned [here](https://www.reddit.com/r/i3wm/comments/fxu7n8/is_it_possible_to_use_bindsym_with_both_release/) and [here](https://www.reddit.com/r/i3wm/comments/hhbbn6/i3wm_ignores_wholewindow_flag_if_released_is/).

If I attempt to implement the same thing with `xbindkeys`, it works:
```scheme
(xbindkey          '(        "m:64" "b:2") "i3-msg kill")
(xbindkey-function '(release "m:64" "b:2") (lambda () #t))
```

I guess it works there because xbindkeys sets up a global hook?

**Comment by budRich** [comment]
@CyberShadow  this one works for me: `bindsym --whole-window Mod4+button2 kill`  
i3-msg is only useful if you want to execute internal i3 commands in a "external" context, you don't need it in the i3 config.

I have this dream key/mouse binding i have been wanting since i started using linux, but could never get it working. I think xbindkeys is the only program that can do it:  

```
button3+button4 -> ctrl+pageup
button3+button5 -> ctrl+pagedown
button3+button2 -> ctrl+w
```

So button3 (right click) acts as a modifier (ctrl) if it is held and button4/5 (scroll up/down) or button2 pressing middle click. Thus will let you control tabs in almost any application with the mouse. Do you know how to code this in xbindkey. I know it is not really i3 related, but this is the closest i have been to my dream.

**Reply by budRich** [reply]
or i seen know that it pastes into the underlying terminal...
Here is a hackish solution, create a script and put it in your path:  
`~/bin/clicknclear`:  

```
#!/bin/bash

blank=$(mktemp)

xclip -sel default "$blank"
rm "$blank"
i3-msg kill
```

Now the binding: `bindsym --whole-window Mod4+button2 exec --no-startup-id clicknclear`  
will work, the drawback is that it will clear the primary clipboard...

**Reply by budRich** [reply]
Endgame:  
`bindsym --whole-window Mod4+button2 exec --no-startup-id "echo -n | xclip -sel default ; i3-msg kill"`

**Reply by CyberShadow** [reply]
Yeah. Weirdly enough `exec "xdotool mouseup 2 ; i3-msg kill"` doesn't work; I think it's because i3 grants a mouse grab to the window as long as any button is down, which interferes with `xdotool` or binding `--whole-window --release`.

**Reply by budRich** [reply]
you could also use xclip to store the current clipboard in a variable, and restore it after the button is released, i guess that will be very noisy in the i3config but in the script it would work fine..

``` bash
#!/bin/bash
cb=$(xclip -o -sel default)
# -r removes the need for echo -n
echo | xclip -r -sel default
i3-msg kill
sleep .2
echo "$cb" | xclip -r -sel default
```

**Reply by CyberShadow** [reply]
Well, the problem isn't exactly that the clipboard gets pasted when the other window is a terminal; it's that the middle click release event gets through to the other window, when it should also be swallowed by the window manager. The terminal case is just an example; it could be any other application, such as a web browser window, where the middle click could e.g. close a tab.

I wonder if it would be possible to write a small C program which breaks the grab and allows fabricating a "mouse up" event that will work.

**Reply by budRich** [reply]
true. Promote this to an issue. Either `--whole-window` should always "swallow" the release event. Or `--release` should work with modifiers. or both off course.

**Reply by budRich** [reply]
just for the record. i don't use this personally, since i am fine with middleclicking the window title with:  
`bindsym --release button2 kill`

**Reply by budRich** [reply]
here is something that seems to work, replace the script with this:  

``` bash
#!/bin/bash

xdotool mouseup 2
{ sleep .1 ; i3-msg kill ;} &
```

**Reply by CyberShadow** [reply]
I think that's the same as `sleep .1 ; i3-msg kill`, i.e., delay the kill so that the mouse up goes to the window that's about to be killed too. It works only if you hold the mouse button down for less than 100ms.

**Reply by budRich** [reply]
I` miss AutoHotkey so much right now ;;
https://gist.github.com/budRich/6044613

**Reply by budRich** [reply]
i haven't tested it myself, but in this issue there is a patch to i3 that might resolve this:
#3459 (Mousebinding with --release and --whole-window doesn't work)

**Reply by CyberShadow** [reply]
Thanks for that link! Some very useful context and details there too. I'm test-driving that patch right now. It does not allow binding `kill` to the mouse down event, but putting `nop`/`kill` on down/up respectively does seem to work reliably.
